### ISSUE TITLE
Rescheduling task on failed host can get stuck on ephemeral disk migration 

### ISSUE THREAD

**Issue by CyBeRoni** [hashicorp/nomad#20587]
### Nomad version
Nomad v1.7.6
BuildDate 2024-03-12T07:27:36Z
Revision 594fedbfbc4f0e532b65e8a69b28ff9403eb822e
### Operating system and Environment details
Debian 11

### Issue

I had a job fail to start after being rescheduled because it was blocked waiting for ephemeral disk migration. In this case, the node failed because its storage failed but the node itself remained reachable on the network and processes kept running (insofar as they didn't crash or fail on their own). 

Nomad did correctly mark the node as failed and correctly rescheduled the job, but also tried to migrate the ephemeral disk. This operation only failed after several hours, when I rebooted the failed node. 

This generated the following logging:

```
May 14 04:31:05 nomad01 nomad[1571988]:     **2024-05-14T04:31:05.483+0200** [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=25557839-5ce5-4df3-5531-9ffc277d19f0 task=vector type=Received msg="Task received by client" failed=false
May 14 11:35:16 nomad01 nomad[1571988]:     **2024-05-14T11:35:16.034+0200** [WARN]  client.alloc_runner.runner_hook.migrate_disk: error migrating data from previous alloc: alloc_id=25557839-5ce5-4df3-5531-9ffc277d19f0 error="error getting snapshot from previous alloc \"d9f1c7be-f979-9cd0-075e-1bcace5f24a6\": Get \"http://192.168.1.103:4646/v1/client/allocation/d9f1c7be-f979-9cd0-075e-1bcace5f24a6/snapshot\": read tcp 192.168.0.17:45488->192.168.1.103:4646: read: connection reset by peer"
```

The node failed and the task was rescheduled at 04:31, and the migration blocked until the affected node was rebooted at 11:35 later that morning.

### Reproduction steps

Have a job with ephemeral disk migration and then kill the disk i/o on the node causing nomad to both remain reachable but also get marked as failed by nomad.

#### Expected Result

 - Jobs that were rescheduled because the node running them failed should probably not attempt to migrate their ephemeral storage. I feel like it can be reasonably safely assumed to also have failed.
 - Where migration _is_ attempted, reasonable network i/o timeouts should exist such that migration does not block without progress for more than a few minutes. 

#### Actual Result

Job blocked for extended periods on data migration.

**Comment by tgross** [comment]
Hi @CyBeRoni! Sorry you hit this issue... this is a really interesting failure mode! If the network connection failed, this would have been stopped because we'd get an EOF. And on the read side on the old host, we can detect _errors_ but we're just never getting a response back from a `read` syscall it's going to be dead in the water.

I took a quick look and it looks like the code that performs the migration (ref [`alloc_watcher.go#L562-L570`](https://github.com/hashicorp/nomad/blob/v1.8.0-beta.1/client/allocwatcher/alloc_watcher.go#L562-L570)) does have a facility to be canceled both at the start and if [we stop receiving chunks](https://github.com/hashicorp/nomad/blob/v1.8.0-beta.1/client/allocwatcher/alloc_watcher.go#L652-L670). But unfortunately that cancellation context is just [`context.TODO()`](https://github.com/hashicorp/nomad/blob/v1.8.0-beta.1/client/allocrunner/migrate_hook.go#L41) which will never be canceled. :facepalm: Looks like that got added years ago during a big client code overhaul and never got addressed.

It's not obvious to me what "reasonable" here is for the total length of time, because the allocdir can be almost arbitrarily large. But I think what we can probably do is something like:
* If we get no response after a short timeout, exit.
* If we start getting a response and then stop getting new chunks after a short timeout, exit.

That timeout can default to something like 10s because if you haven't received anything by then you're probably not getting it. But this also probably wants to be a client configuration tunable.

I'm going to mark this issue for roadmapping.

**Comment by CyBeRoni** [comment]
> It's not obvious to me what "reasonable" here is for the total length of time, because the allocdir can be almost arbitrarily large. 

Yes, my suggestion was to place a timeout on progress; meaning if no data was received for <X> time, assume something is wrong and fail the operation. In this case so far as I can tell it was an HTTP request that went from the new client to the old one that waited for a response for effectively ever, but never got one.

That said it might actually also be useful to have a maximum migration duration, where an operator can specify "if the migration takes longer than <configured time>, abandon ship and start the new task without". This lets me quantify my priorities: the ephemeral data is only important to me if migrating it causes a task/service downtime of less than <configured time>.

**Comment by tgross** [comment]
> That said it might actually also be useful to have a maximum migration duration

Yeah, totally agreed.
